### PR TITLE
VM does not support power opperations if archived

### PR DESCRIPTION
**Pull request by juliancheal**
fixes bz: https://bugzilla.redhat.com/show_bug.cgi?id=1723805

### CODE REVIEW

**Comment by juliancheal** [comment]
@agrare does this look right? I don't have any SCVMM creds to be able to test this.

**Comment by agrare** [comment]
@juliancheal I think what you're looking for is this: https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager/vm/operations/guest.rb#L6

**Comment by agrare** [comment]
In a followup it would be nice if we could do this in an inheritable way, since every VM would have this logic but each provider might tack on additional validations.  Thinking something like move `unsupported_reason_add(:reboot_guest, unsupported_reason(:control)) unless supports_control?` to the core vm model, then here do `super` then the rest of this

File: app/models/manageiq/providers/microsoft/infra_manager/vm.rb [hunk 1]
```diff
@@ -5,15 +5,24 @@ class ManageIQ::Providers::Microsoft::InfraManager::Vm < ManageIQ::Providers::In
   supports_not :publish
 
   supports :reboot_guest do
-    unsupported_reason_add(:reboot_guest, _('The VM is not powered on')) unless current_state == 'on'
+    unsupported_reason_add(:power, unsupported_reason(:control)) unless supports_control?
```
Comment (agrare) [inline app/models/manageiq/providers/microsoft/infra_manager/vm.rb]: This needs to be `unsupported_reason_add(:reboot_guest, ...`

File: app/models/manageiq/providers/microsoft/infra_manager/vm.rb [hunk 2]
```diff
@@ -5,15 +5,24 @@ class ManageIQ::Providers::Microsoft::InfraManager::Vm < ManageIQ::Providers::In
...
+    unless current_state == 'on'
+      unsupported_reason_add(:reboot_guest, _('The VM is not powered on'))
+    end
   end
 
   supports :shutdown_guest do
-    unsupported_reason_add(:shutdown_guest, _('The VM is not powered on')) unless current_state == 'on'
+    unsupported_reason_add(:power, unsupported_reason(:control)) unless supports_control?
```
Comment (agrare) [inline app/models/manageiq/providers/microsoft/infra_manager/vm.rb]: `(:shutdown_guest`

File: app/models/manageiq/providers/microsoft/infra_manager/vm.rb [hunk 3]
```diff
@@ -5,15 +5,24 @@ class ManageIQ::Providers::Microsoft::InfraManager::Vm < ManageIQ::Providers::In
...
+    unless current_state == 'on'
+      unsupported_reason_add(:shutdown_guest, _('The VM is not powered on'))
+    end
   end
 
   supports :reset do
-    unsupported_reason_add(:reset, _('The VM is not powered on')) unless current_state == 'on'
+    unsupported_reason_add(:power, unsupported_reason(:control)) unless supports_control?
```
Comment (agrare) [inline app/models/manageiq/providers/microsoft/infra_manager/vm.rb]: And `(:reset`

**Comment by juliancheal** [comment]
@agrare ahaha copy paste errors thanks!

**Comment by juliancheal** [comment]
@agrare @AlexanderZagaynov and I will make a follow up pr to extract this to base, like you suggested. 👍

**Comment by simaishi** [comment]
Ivanchuk backport details:

```
$ git log -1
commit ec2095bd363f59ede95aa92a00945c1cff776f59
Author: Adam Grare <agrare@redhat.com>
Date:   Wed Jul 24 09:04:26 2019 -0400

    Merge pull request #126 from juliancheal/supports_power_op
    
    VM does not support power opperations if archived
    
    (cherry picked from commit 686039af1deec4f94e15983957cfd665c78b3710)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1723805
```

**Comment by hstastna** [comment]
I've tested that this PR fixes also my BZ to 5.10: https://bugzilla.redhat.com/show_bug.cgi?id=1745915
I am adding appropriate labels

@miq-bot add_label hammer/yes, blocker

**Comment by hstastna** [comment]
@simaishi

**Comment by simaishi** [comment]
`shutdown_guest` and `reboot_guest` don't exist in hammer branch, so backported the changes for `reset` only.

Hammer backport details:

```
$ git log -1
commit 41d314f3564ba17ffdc0081e02c339ac28c4da86
Author: Adam Grare <agrare@redhat.com>
Date:   Wed Jul 24 09:04:26 2019 -0400

    Merge pull request #126 from juliancheal/supports_power_op
    
    VM does not support power opperations if archived
    
    (cherry picked from commit 686039af1deec4f94e15983957cfd665c78b3710)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1748505
```

**Pull request merged**